### ISSUE TITLE
Query with `*` and `.` for global type delarations not work

### ISSUE THREAD

**Issue by shallowclouds** [tree-sitter/tree-sitter#2477]
I want to capture a Golang file's all global type declarations, so I use S-expression query below:

```
(source_file
  (
    (comment)*
    .
    (type_declaration)
  )@type
)
```

In the query:
- `(source_file` indicates the captures must be the direct children of top file(global type declarations, as Go also allow defining types in the function body); 
- `(comment)*` indicates there could be zero or more comment lines for the types;
- `.`  indicates the comment and type declaration should be the immediate siblings(or adjacent nodes).

But it got no captures with this query, are there any mistakes I made or it's just a bug(feature) of treesitter?

## Reproduce

Use the official playground: https://tree-sitter.github.io/tree-sitter/playground: 

<img width="834" alt="image" src="https://github.com/tree-sitter/tree-sitter/assets/15541038/21a84481-83d8-44e6-bbca-8f56a86b1313">

## Other tries

If I remove the `source_file`, it did get captures, but it will also capture type declarations in a function body, which I don't want.

<img width="750" alt="image" src="https://github.com/tree-sitter/tree-sitter/assets/15541038/48965180-0245-42d0-a37f-6c20f3328afa">

<img width="750" alt="image" src="https://github.com/tree-sitter/tree-sitter/assets/15541038/9ed9fdaf-4ea3-4f47-924e-3ea360d2fd8f">

**Comment by amaanq** [comment]
For me it works fine, some sample input code would be useful, instead of just clipped screenshots:

Code:
```go
package main

func a() {
}

// Hello
type b struct{}

func c() {
	type d struct{}
}
```


Output:

<img width="728" alt="image" src="https://github.com/tree-sitter/tree-sitter/assets/29718261/ab895c32-aa90-4ddf-ae6d-08683746f6e5">

**Comment by shallowclouds** [comment]
Yes, for your example, it works fine, but if I add a comment before the package statement(which is common for Go), it stops working again.

```Go
// Package main is my test package.
package main

func a() {
}

// Hello
type b struct{}

func c() {
	type d struct{}
}
```

```
(source_file
  (
    (comment)*
    .
    (type_declaration)
  )@type
)
```

<img width="773" alt="image" src="https://github.com/tree-sitter/tree-sitter/assets/15541038/4c08feeb-b03c-4b36-8866-d05748647c2e">

**Comment by amaanq** [comment]
Yeah I can reproduce that now, thanks!

**Comment by shallowclouds** [comment]
Can we assign a `bug` label for this issue? in case someone may want to fix it.

**Comment by amaanq** [comment]
I believe this is related

https://github.com/tree-sitter/tree-sitter/issues/2514

**Comment by ahlinc** [comment]
> I believe this is related
> 
> #2514

But the above queries don't use predicates at all.
